File: composeApp/src/androidMain/kotlin/org/example/project/feature/map_view/MapContent.kt
package org.example.project.feature.map_view

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.FloatingActionButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.IconButtonDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.vectorResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import com.example.compose.DeliveryAppTheme
import org.example.project.R
import org.example.project.domain.models.DeliveryType
import org.example.project.domain.models.DepartmentModel
import org.example.project.feature.ui_components.SelectedButton
import org.example.project.features.map.UiPoint

@Composable
fun MapContent(
    modifier: Modifier = Modifier,
    addressString: String?,
    deliveryInfo: String?,
    isLoading: Boolean,
    isSearching: Boolean,
    isConfirmEnabled: Boolean,
    moveToLocation: Boolean,
    deliveryType: DeliveryType,
    departments: List<DepartmentModel>,
    selectedDepartment: Int? = null,
    location: UiPoint?,
    onSelectDeliveryType: (DeliveryType) -> Unit = {},
    onConfirm: () -> Unit = {},
    onBackButtonClicked: () -> Unit = {},
    onMapMoved: (Double, Double, Int, Boolean) -> Unit = { _, _, _, _ -> },
    onSelectDepartment: (Int) -> Unit = {_ -> },
    onRequestLocation: () -> Unit = {},
    onSearchClicked: () -> Unit = {},
    showBackButton: Boolean = true,
    showMap: Boolean = true,
) {
    val context = LocalContext.current
    if (isLoading) {
        Dialog(
            onDismissRequest = {}
        ) {
            CircularProgressIndicator()
        }
    }

    Box(
        modifier = modifier
    ) {
        if (showMap) {
            YandexMapView(
                modifier = Modifier
                    .fillMaxSize(),
                context = context,
                position = location,
                moveToLocation = moveToLocation,
                onMapMoved = { lat, lon, reason, finished ->
                    onMapMoved(lat, lon, reason, finished)
                },
                showDepartments = deliveryType == DeliveryType.PICKUP,
                selectedDepartment = selectedDepartment,
                departments = departments,
                onSelectDepartment = onSelectDepartment
            )
        }

        Column(
            modifier = Modifier
                .align(Alignment.TopCenter)
                .fillMaxWidth()
                .padding(horizontal = 16.dp)
                .statusBarsPadding(),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth(),
                horizontalArrangement = Arrangement.End,
            ) {
                if (showBackButton) {
                    IconButton(
                        colors = IconButtonDefaults.iconButtonColors(
                            containerColor = MaterialTheme.colorScheme.secondaryContainer,
                            contentColor = MaterialTheme.colorScheme.onSecondaryContainer
                        ),
                        onClick = onBackButtonClicked
                    ) {
                        Icon(
                            imageVector = ImageVector.vectorResource(R.drawable.ic_arrow_back_16),
                            tint = MaterialTheme.colorScheme.primaryContainer,
                            contentDescription = null,
                        )
                    }
                }
                Spacer(modifier = Modifier.weight(1f))
                if (deliveryType == DeliveryType.DELIVERY) {
                    IconButton(
                        colors = IconButtonDefaults.iconButtonColors(
                            containerColor = MaterialTheme.colorScheme.secondaryContainer,
                            contentColor = MaterialTheme.colorScheme.onSecondaryContainer
                        ),
                        onClick = onSearchClicked
                    ) {
                        Icon(
                            imageVector = ImageVector.vectorResource(R.drawable.ic_loupe_16),
                            tint = MaterialTheme.colorScheme.primaryContainer,
                            contentDescription = null,
                        )
                    }
                }
            }
            if (addressString != null) {
                Text(
                    style = MaterialTheme.typography.headlineLarge,
                    textAlign = TextAlign.Center,
                    color = MaterialTheme.colorScheme.onBackground,
                    text = addressString
                )
            }
            if (deliveryInfo != null) {
                Text(
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onBackground,
                    modifier = Modifier
                        .clip(RoundedCornerShape(18.dp))
                        .background(MaterialTheme.colorScheme.secondaryContainer)
                        .padding(vertical = 8.dp, horizontal = 16.dp),
                    text = deliveryInfo,
                )
            }
        }
        if (deliveryType == DeliveryType.DELIVERY) {
            TooltipComponent(
                modifier = Modifier
                    .align(Alignment.Center)
                    .offset(y = (-24).dp),
                isSearching = isSearching,
            )
        }
        Column(
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .fillMaxWidth()
        ) {
            FloatingActionButton(
                modifier = Modifier
                    .align(Alignment.End)
                    .padding(16.dp),
                shape = RoundedCornerShape(50),
                elevation = FloatingActionButtonDefaults.elevation(
                    defaultElevation = 4.dp
                ),
                onClick = {
                    onRequestLocation()
                }
            ) {
                Icon(
                    imageVector = ImageVector.vectorResource(R.drawable.ic_marker_point),
                    contentDescription = null,
                )
            }

            Column(
                modifier = Modifier
                    .background(MaterialTheme.colorScheme.background)
                    .navigationBarsPadding()
                    .padding(16.dp)
            ) {
                Row(
                    modifier = Modifier
                        .padding(bottom = 8.dp)
                        .fillMaxWidth(),
                    horizontalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    SelectedButton(
                        modifier = Modifier
                            .weight(1f),
                        title = "Доставка",
                        selected = deliveryType == DeliveryType.DELIVERY,
                        onClick = {
                            if (deliveryType == DeliveryType.PICKUP) {
                                onSelectDeliveryType(DeliveryType.DELIVERY)
                            }
                        }
                    )
                    SelectedButton(
                        modifier = Modifier
                            .weight(1f),
                        title = "Самовывоз",
                        selected = deliveryType == DeliveryType.PICKUP,
                        onClick = {
                            if (deliveryType == DeliveryType.DELIVERY) {
                                onSelectDeliveryType(DeliveryType.PICKUP)
                            }
                        }
                    )
                }
                Button(
                    modifier = Modifier.fillMaxWidth(),
                    elevation = ButtonDefaults.buttonElevation(
                        defaultElevation = 4.dp
                    ),
                    shape = RoundedCornerShape(50),
                    enabled = isConfirmEnabled,
                    onClick = onConfirm,
                ) {
                    Text(
                        color = MaterialTheme.colorScheme.background,
                        text = "Подтвердить"
                    )
                }
            }
        }
    }
}

@Preview(showSystemUi = true)
@Composable
private fun MapContent_Preview() {
    DeliveryAppTheme {
        MapContent(
            modifier = Modifier.fillMaxSize(),
            addressString = "ул. Богдана Хмельницкого, 2А",
            deliveryInfo = "Доставка 100 ₽",
            isLoading = false,
            isSearching = true,
            moveToLocation = false,
            isConfirmEnabled = false,
            location = null,
            deliveryType = DeliveryType.DELIVERY,
            showMap = false,
            showBackButton = true,
            departments = emptyList()
        )
    }
}

File: iosApp/iosApp/feature/map/MapContent.swift
//
//  MapContent.swift
//  iosApp
//
//  Created by Фиркат Давлетов on 27/10/2025.
//  Copyright © 2025 orgName. All rights reserved.
//
import SwiftUI
import Shared

struct MapContent: View {
    let addressString: String?
    let deliveryInfo: String?
    let isLoading: Bool
    let isSearching: Bool
    let isConfirmEnabled: Bool
    let moveToLocation: Bool
    let deliveryType: DeliveryType
    let departments: Array<DepartmentModel>
    let selectedDepartment: KotlinInt?
    let location: UiPoint?
    let onSelectDeliveryType: (DeliveryType) -> Void
    let onConfirm: () -> Void
    let onBackButtonClicked: () -> Void
    let onSearchAddressClicked: () -> Void
    let onMapMoved: (Double, Double, UInt, Bool) -> Void
    let onSelectDepartment: (Int64) -> Void
    let onRequestLocation: () -> Void
    let showBackButton: Bool
    let showSearchButton: Bool
    let showMap: Bool
    
    var body: some View {
        ZStack {
            if (showMap) {
                YandexMapView(
                    position: location,
                    moveToLocation: moveToLocation,
                    showDepartments: deliveryType == DeliveryType.pickup,
                    selectedDepartment: selectedDepartment,
                    departments: departments,
                    deliveryType: deliveryType,
                    onMapMoved: onMapMoved,
                    onSelectDepartments: onSelectDepartment
                )
                .edgesIgnoringSafeArea(.all)
            }
            VStack {
                HStack {
                    if (showBackButton) {
                        IconButton(
                            systemName: "arrow.backward",
                            tint: Color.primaryContainer,
                            foreground: Color.onPrimaryContainer,
                            action: onBackButtonClicked
                        )
                    }
                    Spacer()
                }
                
                if (addressString != nil) {
                    Text(addressString!)
                        .font(AppTypography.headlineLarge)
                        .foregroundStyle(Color.darkCosmicBlue)
                        .multilineTextAlignment(.center)
                }
                if (deliveryInfo != nil) {
                    Text(deliveryInfo!)
                        .font(AppTypography.bodyMedium)
                        .padding(.horizontal, 16)
                        .padding(.vertical, 8)
                        .background(
                            RoundedRectangle(cornerRadius: 16)
                                .fill(Color.primaryContainer)
                        )
                        .foregroundStyle(Color.onPrimaryContainer)
                }
                Spacer()
            }
            .padding(.horizontal)
            
            if (deliveryType == DeliveryType.delivery) {
                TooltipComponent()
            }
            
            VStack {
                Spacer()
                if (deliveryType == DeliveryType.delivery) {
                    IconButton(
                        systemName: "magnifyingglass",
                        tint: Color.primaryContainer,
                        foreground: Color.onPrimaryContainer,
                        action: onSearchAddressClicked
                    )
                    .frame(maxWidth: .infinity, alignment: .trailing)
                    .padding()
                    .shadow(color: .black.opacity(0.2), radius: 4, x: 0, y: 2)
                }
                
                IconButton(
                    systemName: "location",
                    tint: Color.primaryContainer,
                    foreground: Color.onPrimaryContainer,
                    action: onRequestLocation
                )
                    .frame(maxWidth: .infinity, alignment: .trailing)
                    .padding()
                    .shadow(color: .black.opacity(0.2), radius: 4, x: 0, y: 2)
                VStack {
                    HStack {
                        SelectedButton(
                            title: "Самовывоз",
                            selected: deliveryType == DeliveryType.pickup,
                            action: {
                                onSelectDeliveryType(DeliveryType.pickup)
                            },
                        )
                        SelectedButton(
                            title: "Доставка",
                            selected: deliveryType == DeliveryType.delivery,
                            action: {
                                onSelectDeliveryType(DeliveryType.delivery)
                            },
                        )
                    }
                    PrimaryButton(
                        title: "Подтвердить",
                        onClick: onConfirm,
                        enabled: isConfirmEnabled
                    )
                    .disabled(!isConfirmEnabled)
                }
                .padding(.horizontal)
            }
        }
    }
}

#Preview {
    MapContent(
        addressString: "ул. Косоротова, 56",
        deliveryInfo: "Доставка 100 ₽",
        isLoading: false,
        isSearching: false,
        isConfirmEnabled: true,
        moveToLocation: false,
        deliveryType: DeliveryType.delivery,
        departments: [],
        selectedDepartment: KotlinInt(int: 0),
        location: nil,
        onSelectDeliveryType: { _ in },
        onConfirm: {},
        onBackButtonClicked: {},
        onSearchAddressClicked: {},
        onMapMoved: {_,_,_,_ in },
        onSelectDepartment: {_ in },
        onRequestLocation: {},
        showBackButton: true,
        showSearchButton: true,
        showMap: true
    )
}
